### ISSUE TITLE
Add support for computed indexes

### ISSUE THREAD

**Issue by Fumuran** [0xMiden/air-script#245]
Add support for computed indexes, i.e. `s[i + 1]` which can be used in constraint comprehension.

**Comment by tohrnii** [comment]
We should probably also allow named constants as addends (maybe even variables?).

**Comment by bobbinth** [comment]
One of the reasons why we didn't have this before is that it requires differentiating between scalars which are field elements and scalars which are integers. We might need to think through this in a bit more detail. For example:
```
let x = [i * s[i + 1] for i in 0..5]
```
Here we treat `i` as an integer in `i + 1` expression and as a field element in `i * s[..]` expression.

**Comment by tohrnii** [comment]
Makes sense. Maybe we can allow only very simple expressions for indices and evaluate them before codegen.

**Comment by Leo-Besancon** [comment]
Based on the comments and what I saw, it seems that doing https://github.com/0xPolygonMiden/air-script/issues/334 (Integer / Felt separation) first (set for AirScript v0.5) may make things easier.

What do you think we should do there @bobbinth?

We can maybe either:
- set issue 245 for milestone AirScript v0.5?
- prioritize issue 334 first and set it for AirScript v0.4?
- assume that computed indices can be set from constants, existing ranges (i in 0..5) and binary expressions combining them (so we can deduce i+2 can be used as a computed index), and then improve typing analysis once 334 is done?
- or propagate indices until all MIR passes (Inlining / Unrolling) are done, and allow the computed index if it is a Constant (in this case, as i in 0..5 would be unrolled in a Vec, each element referencing i would be constant). In this case we probably need another constant folding pass.

**Comment by bobbinth** [comment]
I've moved this to v0.5 milestone.
